### PR TITLE
introduce new skipMvnDeploy goal option to support automatical art…

### PR DESCRIPTION
**Pull request by laur89**
…ifact deployments.

  note the option still depends on skipTag, as only tags can be deployed.

The deployment is always done as the last step to ensure broken tags don't get deployed to mvn repo.

`skipMvnDeploy` defaults to `true` as not to break backwards compatibility, but I'd like you to consider opposite default value.

### CODE REVIEW

**Comment by aleksandr-m** [comment]
*as only tags can be deployed.* - Why do you think that?
Overall I was planing to introduce the ability to set profiles and/or to run additional maven goals in various stages of the plugin execution.
But it will take time to implement it, if you really want this deploy feature we can do it now and deprecate later when additional goal running will be ready.

File: README.md [hunk 1]
```diff
@@ -131,8 +131,12 @@ Note that although `${project.version}` can be used, any changes to version intr
 
 ## Additional goal parameters
 
-The `gitflow:release-finish`, `gitflow:release` and `gitflow:hotfix-finish` goals have `skipTag` parameter. This parameter controls whether the release/hotfix will be tagged in Git.
+The `gitflow:release-finish`, `gitflow:release` and `gitflow:hotfix-finish` goals have `skipTag` and `skipMvnDeploy` parameters:
+- `skipTag` controls whether the release/hotfix will be tagged in Git.
 The default value is `false` (i.e. the release/hotfix will be tagged).
+- `skipMvnDeploy` controls whether created Git tag will be deployed to nexus.
+Note deployment also requires `skipTag=true`, since only tags get deployed.
```
Comment (aleksandr-m) [inline README.md]: you mean `skipTag=false`, right?

File: src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java [hunk 2]
```diff
@@ -40,9 +40,9 @@
 
 /**
  * Abstract git flow mojo.
- *
```
Comment (aleksandr-m) [inline src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java]: What is with the formatting? Please remove any stylistic changes.

File: src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java [hunk 3]
```diff
@@ -749,8 +762,25 @@ protected void mvnCleanInstall() throws MojoFailureException,
...
     /**
+     * Checks out given {@code tagVersion}, executes mvn deploy & finally
+     * checks out the previous branch again.
+     *
+     * @throws MojoFailureException
+     * @throws CommandLineException
+     */
+    protected void mvnDeploy(String tagVersion) throws MojoFailureException,
+            CommandLineException {
+        getLog().info( String.format( "Deploying tag '%s' to nexus", tagVersion));
```
Comment (aleksandr-m) [inline src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java]: why nexus?

File: src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java [hunk 4]
```diff
@@ -749,8 +762,25 @@ protected void mvnCleanInstall() throws MojoFailureException,
...
+        final String currentBranch = gitBranchGet();
+
+        gitCheckout(tagVersion);
+        executeMvnCommand("deploy");
+        gitCheckout(currentBranch);
```
Comment (aleksandr-m) [inline src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java]: `mvnDeploy` should only deploy things

File: src/main/java/com/amashchenko/maven/plugin/gitflow/GitFlowHotfixFinishMojo.java [hunk 5]
```diff
@@ -204,6 +214,10 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                     gitPush(gitFlowConfig.getDevelopmentBranch(), !skipTag);
                 }
             }
+
+            if (!skipTag && !skipMvnDeploy) {
+                mvnDeploy(tagVersion);
```
Comment (aleksandr-m) [inline src/main/java/com/amashchenko/maven/plugin/gitflow/GitFlowHotfixFinishMojo.java]: version is not the same thing as branch

**Comment by laur89** [comment]
> as only tags can be deployed

Because of the limitation set by this commit, not mvn. Did that to force tagging the deployment as well, but now that I think about it, it would also force tagging snapshots, which is not something that should be done.

Have you set any ballpark estimation for the planned development?

File: src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java (hunk 2 above)
Comment (laur89) [inline src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java]: IDE had removed trailing whitespace automatically.

File: README.md (hunk 1 above)
Comment (laur89) [inline README.md]: Correct.

File: src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java (hunk 3 above)
Comment (laur89) [inline src/main/java/com/amashchenko/maven/plugin/gitflow/AbstractGitFlowMojo.java]: Yeah you're right; it's just that I've started taking 'nexus' as a synonym to mvn repo.

File: src/main/java/com/amashchenko/maven/plugin/gitflow/GitFlowHotfixFinishMojo.java (hunk 5 above)
Comment (laur89) [inline src/main/java/com/amashchenko/maven/plugin/gitflow/GitFlowHotfixFinishMojo.java]: Not following.

**Comment by mbeniamin** [comment]
Introducing deployment option is really helpful.

**Comment by aleksandr-m** [comment]
@mbeniamin Initial commit is [here](https://github.com/aleksandr-m/gitflow-maven-plugin/commit/1e26c36077e91398638ba7b625810d55dc48aad8). You can test and review.

**Comment by mbeniamin** [comment]
I was able to test and have a look on your feature branch. The support for executing goals before or after release-start / finish is good.
I also reviewed and tested laur89 work that facilitates deployment to artifactory. That should be executed from the release sources that are being tagged I think that from a 'git-flow' process perspective that is more accurate. 
I was able to capture his changes and merge them with your latest version 1.5.0. That is the best I can get and it serves my purpose. 

ps. I love the merge process. This turned my favor from jgit-flow. they got stuck with recursion strategy.

Thanks for your work!

**Comment by aleksandr-m** [comment]
@mbeniamin With the version from feature branch there is no need in this PR. Good to hear some feedback on it.

**Comment by aleksandr-m** [comment]
Implemented in more general form in 9e57bb0.

**Pull request closed without merging**